api/src/api/users/user_schemas.py

```diff
@@ -389,3 +395,130 @@ class UserUpdateProfileRequestSchema(UserProfile):
+
+
+class UserRoleSchema(Schema):
+    """Schema for user role and privileges within a resource."""
+
+    role_id = fields.String(metadata={"description": "The internal ID of a role"})
+    role_name = fields.String(metadata={"description": "The name of the role"})
+    privileges = fields.List(
+        fields.Enum(Privilege, metadata={"description": "Privileges for the role"})
+    )
+
+    @pre_dump
+    def flatten_linked_role(
+        self,
+        resource_role: (
+            ApplicationUserRole | OrganizationUserRole | InternalUserRole | AgencyUserRole
+        ),
+        **kwargs: Any
+    ) -> dict:
+        role = resource_role.role
+        return {
+            "role_id": resource_role.role_id,
+            "role_name": role.role_name,
+            "privileges": role.privileges,
+        }
+
+
+class GetUserRolesAndPrivilegesOrganizationSchema(Schema):
+    organization_id = fields.String(
+        metadata={
+            "description": "The internal ID of the organization",
+        }
+    )
+
+
+class GetUserRolesAndPrivilegesApplicationSchema(Schema):
+    application_id = fields.String(
+        metadata={
+            "description": "The internal ID of the application",
+        }
+    )
+
+
+class GetUserRolesAndPrivilegesAgencySchema(Schema):
+    agency_id = fields.String(
+        metadata={
+            "description": "The internal ID of the agency",
+        }
+    )
+
+
+class UserOrganization(Schema):
+    """Schema representing a user's association with an organization and their roles."""
+
+    organization = fields.Nested(
+        GetUserRolesAndPrivilegesOrganizationSchema,
+        metadata={"description": "Organization details"},
+    )
+    organization_user_roles = fields.List(
+        fields.Nested(
+            UserRoleSchema,
+            metadata={"description": "Roles the user has for this organization"},
+        )
+    )
+
+
+class UserApplication(Schema):
+    """Schema representing a user's association with an application and their roles."""
+
+    application = fields.Nested(
+        GetUserRolesAndPrivilegesApplicationSchema, metadata={"description": "Application details"}
+    )
+    application_user_roles = fields.List(
+        fields.Nested(
+            UserRoleSchema,
+            metadata={"description": "Roles the user has for this application"},
+        )
+    )
+
+
+class UserAgency(Schema):
+    """Schema representing a user's association with an agency and their roles."""
+
+    agency = fields.Nested(
+        GetUserRolesAndPrivilegesAgencySchema, metadata={"description": "Agency details"}
+    )
+    agency_user_roles = fields.List(
+        fields.Nested(
+            UserRoleSchema, metadata={"description": "Roles the user has for this agency"}
+        )
+    )
+
+
+class GetRolesAndPrivilegesResponseSchema(Schema):
+    """Schema for the full response containing user roles and privileges."""
+
+    user_id = fields.UUID()
+    organizations = fields.List(
```

Comment:
I know this just matches what the DB model has setup, but could we rename this `organization_users` to match the others? Ideally we'd rename that relationship, but if that's more than like 5 minutes of work to do, we can add a property like this to the user model for now.

```py
def organization_users(self) -> list[OrganizationUser]:
     return self.organizations
```

Reply:
[peanut gallery]: aren't these all pointing to something more like `organization_user_roles` / `application_user_roles`? As it stands these sound more like they are pointing to the a relationship between the resource and the user rather than to the roles and underlying privileges